mlflow/pyfunc/__init__.py

```diff
@@ -320,62 +452,246 @@ def predict(*args):
+                       ``parameters`` dictionary::
+
+                         {
+                             "my_list": range(10)
+                         }
+
+                       The ``model_class`` can refer to the Python list named ``"my_list"`` as
+                       ``self.context.parameters["my_list"]``.
+
+                       If *None*, no Python object parameters will be added to the model.
     """
-    if os.path.exists(dst_path):
-        raise Exception("Path '{}' already exists".format(dst_path))
-    os.makedirs(dst_path)
-    code = None
-    data = None
-    env = None
-
-    if data_path:
-        model_file = _copy_file_or_tree(src=data_path, dst=dst_path, dst_dir="data")
-        data = model_file
-
-    if code_path:
-        for path in code_path:
-            _copy_file_or_tree(src=path, dst=dst_path, dst_dir="code")
-        code = "code"
-
-    if conda_env:
-        shutil.copy(src=conda_env, dst=os.path.join(dst_path, "mlflow_env.yml"))
-        env = "mlflow_env.yml"
+    first_argument_set = {
+        "loader_module": loader_module,
+        "data_path": data_path,
+    }
+    second_argument_set = {
+        "artifacts": artifacts,
+        "parameters": parameters,
+        "model_class": model_class,
+    }
+    first_argument_set_specified = any([item is not None for item in first_argument_set.values()])
+    second_argument_set_specified = any([item is not None for item in second_argument_set.values()])
+    if first_argument_set_specified and second_argument_set_specified:
+        raise MlflowException(
+            message=(
+                "The following sets of arguments cannot be specified together: {first_set_keys}"
+                " and {second_set_keys}. All arguments in one set must be `None`. Instead, found"
+                " the following values: {first_set_entries} and {second_set_entries}".format(
+                    first_set_keys=first_argument_set.keys(),
+                    second_set_keys=second_argument_set.keys(),
+                    first_set_entries=first_argument_set,
+                    second_set_entries=second_argument_set)),
+            error_code=INVALID_PARAMETER_VALUE)
 
-    add_to_model(model, loader_module=loader_module, code=code, data=data, env=env)
-    model.save(os.path.join(dst_path, 'MLmodel'))
-    return model
+    if loader_module is not None:
+        return mlflow.pyfunc.model._save_model_with_loader_module_and_data_path(
+            path=dst_path, loader_module=loader_module, data_path=data_path,
+            code_paths=code_path, conda_env=conda_env, mlflow_model=model)
+    elif model_class is not None:
+        return mlflow.pyfunc.model._save_model_with_class_artifacts_params(
+            path=dst_path, model_class=model_class, artifacts=artifacts, parameters=parameters,
+            conda_env=conda_env, code_paths=code_path, mlflow_model=model)
+    elif data_path is not None:
```

Comment:
Does it make sense to these error checks before execution on `save_model_1` vs `save_model_2` recipe...

Reply:
Definitely! I've modified the logic to perform parameter validation in each internal method.